libopae/src/error.c

```diff
@@ -0,0 +1,328 @@
+	while (p) {
+		if (i == error_num) {
+			if (!p->info.can_clear) {
+				FPGA_MSG("can't clear error '%s'", p->info.name);
+				return FPGA_NOT_SUPPORTED;
+			}
+			// read current error value
+			res = fpgaReadError(token, error_num, &value);
+			if (res != FPGA_OK)
+				return res;
+
+			// write to 'clear' file
+			if (stat(p->clear_file, &st) == -1) {
+				FPGA_MSG("can't stat %s", p->clear_file);
+				return FPGA_EXCEPTION;
+			}
+			res = sysfs_write_u64(p->clear_file, value);
+			if (res != FPGA_OK) {
+				FPGA_MSG("can't write clear file '%s'", p->clear_file);
+				return res;
+			}
+			return FPGA_OK;
+		}
+		i++;
+		p = p->next;
+	}
+
+	FPGA_MSG("error info %d not found", error_num);
+	return FPGA_NOT_FOUND;
+}
+
+fpga_result fpgaClearAllErrors(fpga_token token)
+{
+	struct _fpga_token *_token = (struct _fpga_token *)token;
+	uint32_t i = 0;
+	fpga_result res = FPGA_OK;
+
+	ASSERT_NOT_NULL(token);
+	if (_token->magic != FPGA_TOKEN_MAGIC) {
+		FPGA_MSG("Invalid token");
+		return FPGA_INVALID_PARAM;
+	}
+
+	struct error_list *p = _token->errors;
+	while (p) {
+		// if error can be cleared
+		if (p->info.can_clear) {
+			// clear error
+			res = fpgaClearError(token, i);
+			if (res != FPGA_OK)
+				return res;
+		}
+		i++;
+		p = p->next;
+	}
+
+	return FPGA_OK;
+}
+
+fpga_result fpgaGetErrorInfo(fpga_token token,
+			     uint32_t error_num,
+			     struct fpga_error_info *error_info)
+{
+	struct _fpga_token *_token = (struct _fpga_token *)token;
+	uint32_t i = 0;
+
+	if (!error_info) {
+		FPGA_MSG("error_info is NULL");
+		return FPGA_INVALID_PARAM;
+	}
+
+	ASSERT_NOT_NULL(token);
+	if (_token->magic != FPGA_TOKEN_MAGIC) {
+		FPGA_MSG("Invalid token");
+		return FPGA_INVALID_PARAM;
+	}
+
+	struct error_list *p = _token->errors;
+	while (p) {
+		if (i == error_num) {
+			memcpy_s(error_info, sizeof(struct fpga_error_info), &p->info, sizeof(struct fpga_error_info));
+			return FPGA_OK;
+		}
+		i++;
+		p = p->next;
+	}
+
+	FPGA_MSG("error info %d not found", error_num);
+	return FPGA_NOT_FOUND;
+}
+
+/* files and directories to ignore when looking for errors */
+const char *errors_exclude[] = {
+	"revision",
+	"uevent",
+	"power",
+	"clear"
+};
+#define NUM_ERRORS_EXCLUDE (sizeof(errors_exclude) / sizeof(char *))
+
+/* files that can be cleared by writing their value to them */
+const char *errors_clearable[] = {
+	"pcie0_errors",
+	"pcie1_errors",
+	"warning_errors",
+	"inject_error"
+};
+#define NUM_ERRORS_CLEARABLE (sizeof(errors_clearable) / sizeof(char *))
+
+/* Walks the given directory and adds error entries to `list`.
+ * This function is called during enumeration when adding tokens to
+ * the global tokens list. When tokens are cloned, their error
+ * lists are only shallowly copied (which works because errors of
+ * a token never change).
+ * returns the number of error entries added to `list` */
+uint32_t __FIXME_MAKE_VISIBLE__
+build_error_list(const char *path, struct error_list **list)
```

Comment:
Just out of curiosity, what happens if we call this twice with the same list? Will it duplicate the list?

Reply:
Yes, it would append all errors it finds in `path` to the list `list`. It actually calls itself recursively. I'll add a comment.